### PR TITLE
Add support to install or remove flows in the switches through events

### LINKED ISSUE #104: Add support to install Flows through Kytos Events
Today, `flow_manager` is only supported for installing flows using REST endpoints. Because of this, it is necessary to implement support to install `flows` using Kytos Events.

### PR DESCRIPTION
**Pull request by cmagnobarbosa**
### :octocat: Are you working on some issue? Identify the issue!

Fix #104 

### :bookmark_tabs: Description of the Change

Add support to install and remove flows in the switches by `dpid` using kytos Events 
(`kytos.flow_manager.flows.(install|delete)`).

This pull request may allow replacement of the REST endpoint `of_lldp` NApp:
- https://github.com/kytos/of_lldp/blob/master/main.py#L135
- https://github.com/kytos/of_lldp/blob/master/main.py#L137

Suggestions of the changes after this pull request
------------------------------------------------------------------

Suggestion to replace the [#L135](https://github.com/kytos/of_lldp/blob/master/main.py#L135):

``` python
event_install = KytosEvent(name='kytos.flow_manager.flows.install',
                           content={'dpid': destination,
                           'flow_dict': data})
self.controller.buffers.app.put(event_install)
```

Replacing [L#137](https://github.com/kytos/of_lldp/blob/master/main.py#L137):

``` python
event_delete = KytosEvent(name='kytos.flow_manager.flows.delete',
                           content={'dpid': destination,
                           'flow_dict': data})
self.controller.buffers.app.put(event_delete)
```

### :computer: Verification Process

- Unit tests are still passing
- Manually checked.

### :page_facing_up: Release Notes

- Add support for install or remove flows in the switches using Kytos Events.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -252,6 +252,21 @@ def list(self, dpid=None):
...
+        Install or delete Flow from on switch identified by DPID.
+        """
+        try:
+            dpid = event.content['dpid']
+            flow_dict = event.content['flow_dict']
+            command = event.content['command']
+            switch = self.controller.get_switch_by_dpid(dpid)
+            self._install_flows(command, flow_dict, [switch])
+        except (KeyError, InvalidCommandError) as error:
+            log.error(f"Error to install flow: {error}")
```
Comment (hdiogenes) [inline main.py:268]: Break the try/except in two, one for each exception.

**Review by hdiogenes — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (cmagnobarbosa) [reply main.py]: Resolved in https://github.com/kytos/flow_manager/pull/129/commits/dfc4ba126992579fc11020a1291cfcea583f92bc

File: main.py [hunk 2]
```diff
@@ -252,6 +252,32 @@ def list(self, dpid=None):
...
+            flow_dict = event.content['flow_dict']
+        except KeyError as error:
+            log.error("Error getting fields to install or remove "
+                      f"Flows: {error}")
+            return
+
+        if event.name == 'kytos.flow_manager.flows.install':
+            command = 'add'
+        else:
+            command = 'delete'
```
Comment (hdiogenes) [inline main.py:272]: Having the `delete` command as a default may lead to hard-to-find bugs in the future. I'd prefer to make it explicit:

```suggestion
        elif event.name == 'kytos.flow_manager.flows.delete':
            command = 'delete'
        else:
            msg = f'Invalid event "{event.name}", should be install|delete'
            raise ValueError(msg)
```

File: main.py [hunk 3]
```diff
@@ -252,6 +252,32 @@ def list(self, dpid=None):
...
+
+        try:
+            switch = self.controller.get_switch_by_dpid(dpid)
+            self._install_flows(command, flow_dict, [switch])
+        except InvalidCommandError as error:
```
Comment (hdiogenes) [inline main.py:277]: Only the code that raise the exception should be inside the try/except.

```suggestion
        switch = self.controller.get_switch_by_dpid(dpid)
        try:
            self._install_flows(command, flow_dict, [switch])
        except InvalidCommandError as error:
```

File: main.py [hunk 4]
```diff
@@ -252,6 +252,32 @@ def list(self, dpid=None):
 
         return jsonify(switch_flows)
 
+    @listen_to('kytos.flow_manager.flows.(install|delete)')
+    def event_add_flow(self, event):
```
Comment (hdiogenes) [inline main.py:256]: Please use a name that better reflects what this method can do.

```suggestion
    def event_flows_install_delete(self, event):
```

File: tests/unit/test_main.py [hunk 5]
```diff
@@ -187,6 +187,33 @@ def test_install_flows_with_delete_strict(self, *args):
...
+        dpid = "00:00:00:00:00:00:00:01"
+        switch = get_switch_mock(dpid)
+        self.napp.controller.switches = {dpid: switch}
+        mock_flow_dict = MagicMock()
+        event = get_kytos_event_mock(name='kytos.flow_manager.flows.delete',
+                                     content={'dpid': dpid,
+                                              'flow_dict': mock_flow_dict})
+        self.napp.event_add_flow(event)
+        mock_install_flows.assert_called_with('delete', mock_flow_dict,
+                                              [switch])
```
Comment (hdiogenes) [inline tests/unit/test_main.py:215]: Please modify the test names and method calls to reflect the method rename suggested above.

**Review by hdiogenes — CHANGES_REQUESTED** [review]

File: tests/unit/test_main.py (hunk 5 above)
Comment (cmagnobarbosa) [reply tests/unit/test_main.py:215]: Resolved in https://github.com/kytos/flow_manager/pull/129/commits/6876bbce163dce2ebdf3ba9f53c06efaf3b73e33

**Pull request merged**